[ONNX] Fix scalar elements in op.Concat

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98509
* #97494

op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

titaiwangms committed:
[ONNX] Fix scalar elements in op.Conca

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -514,11 +514,7 @@ def forward(self, x):
-        "2. [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while "
-        "running Concat node. Name:'Concat_203' Status Message: concat.cc:104 "
-        "PrepareForCompute Cannot concatenate scalars"
+        "flaky test: https://github.com/microsoft/onnx-script/issues/523. Fixed in ORT==1.15"
```
Do you know if there is a ort-nightly version we could pin in CI?

Consider adding a skip based on ORT version too.

titaiwangms replied:
It seems we can get it from AzureDevOps, and as long as AzureDevOps doesn't crash, we should be good. But I feel risky...

titaiwangms replied:
Added skip_unsupported_ort_min_version. @justinchuby LMK if I use packaging.version.parse correctly!

justinchuby replied:
packaging.version lgtm!

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

justinchuby commented on test/onnx/pytorch_test_common.py:
```diff
@@ -157,6 +158,24 @@ def wrapper(self, *args, **kwargs):
     return skip_dec
 
 
+def skip_unsupported_min_ort_version(min_opset_version: str):
```
```suggestion
def skip_unsupported_min_ort_version(reason: str, min_opset_version: str):
```

We should still specify reason

justinchuby replied:
opset version was a typo?

titaiwangms replied:
Done

justinchuby commented on test/onnx/pytorch_test_common.py:
```diff
@@ -157,6 +158,24 @@ def wrapper(self, *args, **kwargs):
+                < packaging.version.parse(min_opset_version).release
+            ):
+                raise unittest.SkipTest(
+                    f"Unsupported onnxruntime version: {self.ort_version} < {min_opset_version}"
```
I would make it clear on what is not supported. This sentence can be read both ways (too new or too old) Maybe something like "ONNX Runtime version ... is older than required version ..." ?

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

justinchuby commented on test/onnx/pytorch_test_common.py:
```diff
@@ -158,16 +158,17 @@ def wrapper(self, *args, **kwargs):
                 raise unittest.SkipTest(
-                    f"Unsupported onnxruntime version: {self.ort_version} < {min_opset_version}"
+                    f"ONNX Runtime version: {version} is older than required version {version}."
+                    f"This reason we skip it: {reason}."
```
```suggestion
                    f"Reason: {reason}."
```

for simplicity

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix scalar elements in op.Concat"


op.Concat wrongly concatenated scalar int, and it would raise errors in ORT. However, we didn't see this bug until SegFault was fixed.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
